checker/raw_result.go

```diff
@@ -125,10 +125,35 @@ type MaintainedData struct {
 	ArchivedStatus       ArchivedStatus
 }
 
+type LicenseAttributionType string
+
+const (
+	// forms of security policy hints being evaluated.
+	LicenseAttributionTypeOther     LicenseAttributionType = "otherAttribution"
+	LicenseAttributionTypeRepo      LicenseAttributionType = "repoAttribution"
```

Comment:
do you have a sense of whether the GH API always outperform our current code? Ie, if we look at the weekly cron results, are there cases where out current code finds a license that the GH APIs do not?

I'm trying to understand whether we should just switch to the GH API entirely. I suppose it's good to have a backup if the platform does not have an API (GitLab, BitBucket, etc), so probably better to keep it.

Reply:
On speed: I think the answer is absolutely yes. It appears that GH establishes the repo's license when a license file is added. In my trials, I kept changing my license and GH immediately gave me my license match in the right hand side of the webGUI and the `curl` would immediate (on the next invocation) show that same answer. So GH cooks the License at the time of the commit (it would appear).

My favorite test is with `torvalds/linux` (it breaks a lot of tools), when I use the GH API I got an answer back (using my changes in `scorecard` in < 5 seconds. When I permitted the legacy `scorecard` code to run, it took 1 min and 10'ish seconds. I am still building my regression suite for this code (to ensure I don't break scoring (unless we gradiate the score) - but when GH does not return a `404` it finishes quickly.

On **misses**: Now, one site I have been testing with is `github.com/twardoch/varfonts-ofl` - GH `404` this site, but the legacy code then exams ALL the files (**this is the time sink overall for the legacy code**) in the repo until it hits/misses and eventually come up with this (I think in this specific case this is a `false positive`):

```
time XXSCORECARD_EXPERIMENTAL=false SCORECARD_V6=true GITHUB_AUTH_TOKEN=${GITHUB_AUTH_TOKEN} /home/vagrant/src/scorecard/scorecard  --show-details --repo=github.com/twardoch/varfonts-ofl --checks=License 
Starting [License]
Finished [License]

RESULTS
-------
Aggregate score: 10.0 / 10

Check scores:
|---------|---------|-----------------------|----------------------------------------------------|--------------------------------------------------------------------------------------------------------|
|  SCORE  |  NAME   |        REASON         |                      DETAILS                       |                                       DOCUMENTATION/REMEDIATION                                        |
|---------|---------|-----------------------|----------------------------------------------------|--------------------------------------------------------------------------------------------------------|
| 10 / 10 | License | license file detected | Info: :                                            | https://github.com/ossf/scorecard/blob/6f84cb92bb88aef0db98813f127f3981f74775c4/docs/checks.md#license |
|         |         |                       | VotoSerifGX-OFL/Fonts/VotoSerifGX-VarTTF/OFL.txt:1 |                                                                                                        |
|---------|---------|-----------------------|----------------------------------------------------|--------------------------------------------------------------------------------------------------------|

real	0m16.701s
user	0m3.862s
sys	0m3.884s
```

By `false positive` I mean that in the reported/found file this text is there `This Font Software is licensed under the SIL Open Font License, Version 1.1.` but in context I don't this it applies to the entire repo `github.com/twardoch/varfonts-ofl`.

I agree with you... by trying the GH (or generally the repo's API (if they have one)) is the best to try first (it is _fast_ for sure), but leave the legacy code in the event that GH (or the repo) fails... the legacy _backup_ would be **no worse** than what it already does.. a best guess.

Reply:
follow-up question: do you think we should *only* use the GH API when running the check on GitHub? I wonder whether scorecard result should be consistent with what GitHub displays in their UI. I wonder whether if we ever provide a different result, some uses may get confused - I doubt this would happen very often though

Reply:
On this I'd like to reserve judgement and make the call before this is merged. I have about 400 sites I was using for a previous PR and I believe I can use those results (going months back) to compare license results with these proposed changes -- I'd like to build a bit more confidence in the GH API use in SC to make a recommendation.

Reply:
Here are results from 372 sites scanned on github using the github API (GH API) to retrieve the license for a repo as compared to using the existing scorecard (SC legacy) code for license detection:

Of the 372 sites scanned

- 87.4% of the sites, GH API and SC legacy were in `Agreement`, that is that both GH API and SC legacy code either found the same file or found NO license file
- 10.2% of the sites, GH API and SC legacy found `Different` files
-- example (1 of 38): for `github.com/eclipse/ditto` SC identified `.github/workflows/license-check.yml` as a license file while GH API identified `LICENSE` 
-- example (2 of 38): for `github.com/apache/iceberg` SC identified `.baseline/copyright/001_apache-2.0.txt` as a license file while GH API identified `LICENSE`
-- example (3 of 38): for `github.com/eclipse/kitalpha` SC identified `Docgen/business/features/org.polarsys.kitalpha.doc.gen.business.core.feature/license.html` as a license file while GH API identified `LICENSE.md`
- 2.4% of the sites, SC legacy identified a file as a license file (arguably a `false positive` on inspection of the identified file) where GH API returned NO license file
-- example (1 of 9): for `github.com/apache/doris-website` SC identified `static/fonts/OFL.txt` as a license file (probably is not for the entire repo)
-- example (2 of 9): for `github.com/eclipse/emf.egf` SC identified `deprecated/3.3/features/org.eclipse.egf.portfolio.genchain.ecoretools.feature/license.html` as a license file (probably is not for the entire repo)

The SC legacy approach to walk the depths of the repo in search of a file is likely the cause of a number of `false positives` (that 2.4%) (although it is often the case that when looking at actual source code, once often finds a license disclaimer at the top). But that same strategy is also the cause of the `Different` files but coming up with the same **license file detected** in SC--just good luck.

My conclusion is two-fold:

1. For repos that support a license API retrieval/query for a repo, like the GH API, use it... period. Don't walk the depths of the repo when the GH API returns **license file not detected**. The documentation and remedy should be clear and follow the site recommendations on establishing a license for the repo.
2. For repo sites where there is not license API like GH (e.g., localdir, and other "forges"), and extensive walk will continue to bring `false positives` if nothing is done to limit the depths of the search. I suspect that such may also be true for other SC checks the walk repos (like Security-Policy).

I can provide more details if necessary

Reply:
I have made the code follow the two-fold conclusion above. It could be possible for the code to “fall” into the old way on GH if the GH API comes back with no license’s found, but that decision could be made at anytime.